File: 2022/python/src/solutions/day_05.py
from typing import NamedTuple

from src.lib.parsing import read_text
from src.lib.solution import SolutionABC


class Move(NamedTuple):
    start: int
    end: int
    number: int


class Solution(SolutionABC):
    @staticmethod
    def common_parse() -> (list[list[str]], list[Move]):
        stack, movements = read_text(5).split("\n\n")
        stacks: list[list[str]] = [[] for _ in stack.splitlines()]
        for line in stack.splitlines()[:-1][::-1]:  # Skipping first line and reversing order
            for ind in range(0, len(line), 4):  # Each group is 4 characters wide
                group = line[ind: ind + 4]
                if group[1] != " ":
                    stacks[ind // 4].append(group[1])

        moves: list[Move] = []
        for line in movements.splitlines():  # eg: move 5 from 4 to 5
            numbers = line.split()
            number = int(numbers[1])  # eg: 5
            start = int(numbers[3]) - 1  # eg: 4  || Converting 1-index from input to 0-index
            end = int(numbers[5]) - 1  # eg: 5  || Converting 1-index from input to 0-index
            moves.append(Move(start, end, number))

        return stacks, moves

    def common_solve(self, reverse: bool) -> str:
        stacks, moves = self.common_parse()
        for move in moves:
            work_load = []
            for _ in range(move.number):
                crate = stacks[move.start].pop()
                work_load.append(crate)

            work_load = work_load[::-1] if reverse else work_load
            stacks[move.end] += work_load

        result = "".join([stack[-1] for stack in stacks])

        return result

    def _part_a(self) -> str:
        return self.common_solve(reverse=False)

    def _part_b(self) -> str:
        return self.common_solve(reverse=True)

File: 2022/rust/src/solution/day_05.rs
use crate::common::io::read_day;
use crate::common::solution::Solution;


enum Order {
    Unchanged,
    Reversed
}

#[derive(Clone, Debug)]
struct Move {
    number: usize,
    start: usize,
    end: usize,
}

pub struct Day {
    moves: Vec<Move>,
    stacks: Vec<Vec<char>>,
}

impl Day {
    pub fn new() -> Self {
        let input = read_day(5).unwrap();
        let (stack, movements) = input.split_once("\n\n").unwrap();
        let mut stacks: Vec<Vec<char>> = vec![Vec::new(); 9];
        for line in stack.lines().rev().skip(1) {
            for ind in (0..line.len()).step_by(4) {
                if let Some(char) = line.chars().nth(ind + 1) {
                    if char.is_alphabetic() {
                        stacks[ind / 4].push(char)
                    }
                }
            }
        }

        let moves: Vec<Move> = movements
            .lines()
            .map(|line| line
                .split(' ')
                .filter_map(|num| num.parse::<usize>().ok())
            )
            .map(|mut nums| {
                Move {
                    number: nums.next().unwrap(),
                    start: nums.next().unwrap() - 1,
                    end: nums.next().unwrap() - 1,
                }
            }
            )
            .collect();

        Day {moves, stacks}
    }

    fn solve(&self, order: Order) -> String {
        let mut stacks = self.stacks.clone();
        for mov in &self.moves {
            let ind = stacks[mov.start].len() - mov.number;
            let workload: Vec<char> = stacks[mov.start].drain(ind..).collect();
            match order {
                Order::Unchanged => stacks[mov.end].extend(workload),
                Order::Reversed => stacks[mov.end].extend(workload.iter().rev())
            }
        }

        stacks.iter_mut().map(|stack| stack.pop().unwrap()).collect()
    }
}


impl Solution for Day {
    fn name(&self) -> &'static str { "Day 5: Supply Stacks" }

    fn part_a(&self) -> Option<String> {
        Some(self.solve(Order::Reversed))  // RLFNRTNFB
    }

    fn part_b(&self) -> Option<String> {
        Some(self.solve(Order::Unchanged))  // MHQTLJRLB
    }
}

